File: CPP/MaximumGap.cpp
#include <algorithm>
#include <iostream>
#include <optional>
#include <vector>

class Solution {
public:
    int maximumGap(std::vector<int>& nums) {
        int n = nums.size();
        if (n < 2) return 0;
        auto [mnIt, mxIt] = std::minmax_element(nums.begin(), nums.end());
        int mn = *mnIt, mx = *mxIt;
        if (mn == mx) return 0;

        // Pigeonhole: max gap >= ceil((mx-mn)/(n-1)), so it lies between buckets of that size.
        int bucketSize = std::max(1, (mx - mn) / (n - 1));
        int numBuckets = (mx - mn) / bucketSize + 1;
        std::vector<std::optional<std::pair<int, int>>> buckets(numBuckets);

        for (int x : nums) {
            int i = (x - mn) / bucketSize;
            if (!buckets[i]) buckets[i] = {x, x};
            else {
                buckets[i]->first = std::min(buckets[i]->first, x);
                buckets[i]->second = std::max(buckets[i]->second, x);
            }
        }

        int best = 0, prevMax = mn;
        for (auto& b : buckets) {
            if (!b) continue;
            best = std::max(best, b->first - prevMax);
            prevMax = b->second;
        }
        return best;
    }
};

int main() {
    Solution s;
    std::vector<int> a = {3, 6, 9, 1};
    std::vector<int> b = {10};
    std::vector<int> c = {1, 1, 1, 1};
    std::vector<int> d = {1, 10000000};
    std::cout << s.maximumGap(a) << "\n"; // 3
    std::cout << s.maximumGap(b) << "\n"; // 0
    std::cout << s.maximumGap(c) << "\n"; // 0
    std::cout << s.maximumGap(d) << "\n"; // 9999999
    return 0;
}

File: Rust/src/bin/maximum_gap.rs
fn maximum_gap(nums: Vec<i32>) -> i32 {
    let n = nums.len();
    if n < 2 {
        return 0;
    }
    let (mn, mx) = nums.iter().fold((nums[0], nums[0]), |(lo, hi), &x| {
        (lo.min(x), hi.max(x))
    });
    if mn == mx {
        return 0;
    }

    // Pigeonhole: max gap >= ceil((mx-mn)/(n-1)), so it lies between buckets of that size.
    let bucket_size = ((mx - mn) as usize / (n - 1)).max(1);
    let num_buckets = (mx - mn) as usize / bucket_size + 1;
    let mut buckets: Vec<Option<(i32, i32)>> = vec![None; num_buckets];

    for &x in &nums {
        let i = (x - mn) as usize / bucket_size;
        buckets[i] = Some(match buckets[i] {
            None => (x, x),
            Some((lo, hi)) => (lo.min(x), hi.max(x)),
        });
    }

    let mut best = 0;
    let mut prev_max = mn;
    for b in buckets.into_iter().flatten() {
        let (lo, hi) = b;
        best = best.max(lo - prev_max);
        prev_max = hi;
    }
    best
}

fn main() {
    println!("{}", maximum_gap(vec![3, 6, 9, 1]));  // 3
    println!("{}", maximum_gap(vec![10]));          // 0
    println!("{}", maximum_gap(vec![1, 1, 1, 1]));  // 0
    println!("{}", maximum_gap(vec![1, 10000000])); // 9999999
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn examples() {
        assert_eq!(maximum_gap(vec![3, 6, 9, 1]), 3);
        assert_eq!(maximum_gap(vec![10]), 0);
        assert_eq!(maximum_gap(vec![1, 1, 1, 1]), 0);
        assert_eq!(maximum_gap(vec![1, 10_000_000]), 9_999_999);
    }
}
